File: ObjectiveCDemo20200321/ObjectiveCDemo20200321/SHBaseTableViewController.m
//
//  SHBaseTableViewController.m
//  ObjectiveCDemo20200321
//
//  Created by Mac on 2020/4/17.
//  Copyright © 2020 远平. All rights reserved.
//

#import "SHBaseTableViewController.h"

@interface SHBaseTableViewController ()<UITableViewDelegate,UITableViewDataSource>


@end

@implementation SHBaseTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.dataArray = @[].mutableCopy;
    self.sectionTitle = @[].mutableCopy;
    
    self.type = SHBaseTableTypeNewVC;
    
    self.navigationController.interactivePopGestureRecognizer.delegate = (id)self;
    self.tableView=({
        UITableView *tableView=[UITableView new];
        [self.view addSubview:tableView];
        tableView.frame=CGRectMake(0, 0, CGRectGetWidth(self.view.frame), CGRectGetHeight(self.view.frame));
        tableView.delegate = self;
        tableView.dataSource=self;
        tableView;
    });
}

- (void)addSectionDataWithClassNameArray:(NSArray *)classNameArray titleArray:(NSArray *)titleArray title:(NSString *)title {
    @autoreleasepool {
        NSMutableArray *firstArray=@[].mutableCopy;
        
        [classNameArray enumerateObjectsUsingBlock:^(id  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
            SGHCellModel *cellModel = [SGHCellModel new];
            cellModel.className=obj;
            cellModel.title=titleArray[idx];
            [firstArray addObject:cellModel];
        }];
        
        [self p_addSectionTitle:title dataArray:firstArray];
        
        
    }
}

-(void)p_addSectionTitle:(NSString *)sectionTitle dataArray:(NSMutableArray *)dataArray {
    [self.dataArray addObject:dataArray];
    [self.sectionTitle addObject:sectionTitle];
}


#pragma mark - Table view data source
-(NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return self.dataArray.count;
}
-(CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section {
    return 30.0;
}
-(NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    return self.sectionTitle[section];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return ((NSMutableArray *)self.dataArray[section]).count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"YY"];
    if (!cell) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"YY"];
    }
    SGHCellModel *cellMoel= self.dataArray[indexPath.section][indexPath.row];
    cell.textLabel.text = cellMoel.title;
    cell.textLabel.numberOfLines = 0;
    cell.textLabel.font = [UIFont systemFontOfSize:13];
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    SGHCellModel *cellMoel= self.dataArray[indexPath.section][indexPath.row];
    NSString *className = cellMoel.className;
    
    if (self.type == SHBaseTableTypeNewVC) {
        if ([self.inStoryboardVCArray containsObject:className]) {
            UIViewController *vc = [[UIStoryboard storyboardWithName:@"Main" bundle:nil] instantiateViewControllerWithIdentifier:className];
            vc.title = cellMoel.title;
            vc.hidesBottomBarWhenPushed = YES;
            [self.navigationController pushViewController:vc animated:YES];
        } else {
            Class cls=NSClassFromString(className);
            if (cls) {
                UIViewController *vc = [cls new];
                vc.title = cellMoel.title;
                vc.hidesBottomBarWhenPushed = YES;
                [self.navigationController pushViewController:vc animated:YES];
            }
        }
    } else {
        
        SEL sel = NSSelectorFromString(className);
        [self performSelector:sel];
    }
    
    
    
}

@end

File: RxSwiftDemo2019_03_31/RxSwiftDemo2019_03_31/SHBaseTableViewController.swift
//
//  SHBaseViewController.swift
//  RxSwiftDemo2019_03_31
//
//  Created by Mac on 2020/4/16.
//  Copyright © 2020 远平. All rights reserved.
//

import UIKit

class SGHCellModel: NSObject {
    var title: String = ""
    var className: String = ""
    
    init(title: String, className: String) {
        self.title = title
        self.className = className
    }
    
    override convenience init() {
        self.init(title:"", className: "")
    }
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

enum SHBaseTableActionType {
    case method //执行方法
    case newVC  //跳到新的控制器
}

class SHBaseTableViewController: UIViewController, UIGestureRecognizerDelegate {
    
    public var actionType: SHBaseTableActionType = .newVC
    
    var tableView : UITableView?
    var dataArray = [[SGHCellModel]]()
    var sectionTitle = [String]()
    
    var inStoryboardVCArray: [String]?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        //设置滑动返回
        self.navigationController?.interactivePopGestureRecognizer?.delegate = self
        
        self.tableView = {
            //初始化tableView的数据
            let tableV = UITableView(frame: CGRect.zero, style: .plain)
            //设置tableView的数据源
            tableV.dataSource = self
            //设置tableView的委托
            tableV.delegate = self
            tableV.register(UITableViewCell.self, forCellReuseIdentifier: "bookmarksCell")
            tableV.register(UITableViewHeaderFooterView.self, forHeaderFooterViewReuseIdentifier: "secHeaderIdentifier")
            self.view.addSubview(tableV)
            return tableV
        }()
        self.tableView!.snp.makeConstraints { (make) in
            make.edges.equalToSuperview()
        }
        
    }
    
    public func p_addSectionData(with classNameArray:[String], titleArray:[String],title:String) {
        var firstArray = [SGHCellModel]()
        for (idx,item) in titleArray.enumerated() {
            let cellModel = SGHCellModel()
            cellModel.className = classNameArray[idx]
            cellModel.title = item
            firstArray.append(cellModel)
        }
        
        self.addSection(with: title, sectionArray: firstArray)
    }
    
    private func addSection(with title:String, sectionArray: [SGHCellModel]) {
        self.dataArray.append(sectionArray)
        self.sectionTitle.append(title)
    }
}

extension SHBaseTableViewController : UITableViewDataSource {
    func numberOfSections(in tableView: UITableView) -> Int {
        return self.dataArray.count
    }
    
    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return 50.0
    }

    func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        let tView = tableView.dequeueReusableHeaderFooterView(withIdentifier: "secHeaderIdentifier")
        
        tView?.textLabel?.numberOfLines = 0
        tView?.textLabel?.font = UIFont.systemFont(ofSize: 10)
        tView?.textLabel?.text = self.sectionTitle[section]
        return tView
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.dataArray[section].count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "bookmarksCell", for:indexPath)
        let cellModel = self.dataArray[indexPath.section][indexPath.row]
        cell.textLabel?.text = cellModel.title
        cell.textLabel?.numberOfLines = 0
        cell.textLabel?.font = UIFont.systemFont(ofSize: 13)
        return cell
    }
}

extension SHBaseTableViewController: UITableViewDelegate {
    
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        
        tableView.deselectRow(at: indexPath, animated: true)
        
        let cellModel = self.dataArray[indexPath.section][indexPath.row]
        let className = cellModel.className
        
        ///解决方案来自：[Swift 通过字符串 转换成对应的 UIViewController](https://www.jianshu.com/p/d5daa8485227)
        
        if self.actionType == .newVC {
            // 1.动态获取命名空间
            guard let nameSpace = Bundle.main.infoDictionary!["CFBundleExecutable"]as? String else{
            return;
            }
            // 2.根据字符串获取对应的Class并转成控制器的类型
            if let cls = NSClassFromString(nameSpace + "." + className) as? UIViewController.Type {
                
                if inStoryboardVCArray?.contains(className) ?? false {
                    
                    let vc = UIStoryboard(name: "Bookmark", bundle: Bundle(for: cls)).instantiateViewController(withIdentifier: className)
                    vc.title = cellModel.title
                    vc.hidesBottomBarWhenPushed = true
                    self.navigationController?.pushViewController(vc, animated: true)
                    
                } else {

                    // 3.创建对应的控制器对象
                    let vc: UIViewController = cls.init()
                    vc.title = cellModel.title
                    vc.hidesBottomBarWhenPushed = true
                    self.navigationController?.pushViewController(vc, animated: true)
                }
            }
            
        } else {
            //执行方法
            let sel = NSSelectorFromString(className)
            self.perform(sel)
        }
        
        
    }
}

